### ISSUE TITLE
Data event comes before being able to handle it

### ISSUE THREAD

**Issue by srgl** [JoshGlazebrook/socks#25]
I am developing smtp client with socks support and initial hello packet from the server comes before I am able to attach `data` event handler:
```
connect (host, port) {
    return new Promise((resolve, reject) => {
      let t = setTimeout(() => reject('socket connect timeout'), this.timeout)
      if (this.proxyHost) {
        SocksClient.createConnection(this.socksOptions(host, port), (err, info) => {
          if (err) return reject(err)
          this.client = info.socket
          this.client.on('data', this.dataHandler.bind(this))
          clearTimeout(t)
          resolve()
        })
      } else {
        this.client = new Socket()
        this.client.on('data', this.dataHandler.bind(this))
        this.client.connect(port, host, () => {
          clearTimeout(t)
          resolve()
        })
      }
    })
  }
```
I am using ssh -D as a socks server. Any workaround here?

**Comment by JoshGlazebrook** [comment]
Hmm, can you possibly try with the changes in #26 ?

**Comment by JoshGlazebrook** [comment]
Actually, are you sure the server is sending data? Looking at the SMTP protocol the client initiates with a "HELO" first before the sends anything back?

**Comment by srgl** [comment]
@JoshGlazebrook Yes, client initiates with a HELO geeting, but before that there is a greeting from the server, try:
`nc gmail-smtp-in.l.google.com 25`

**Comment by JoshGlazebrook** [comment]
Hmm, just tried it myself using a local socks server and I'm seeing it being receiving with my .on('data') handler.

```javascript
const options: SocksClientOptions = {
  proxy: {
    ipaddress: '127.0.0.1',
    port: 1080,
    type: 5
  },

  command: 'connect',

  destination: {
    host: 'gmail-smtp-in.l.google.com',
    port: 25
  }
};

async function test() {
  const result = await SocksClient.createConnection(options);
  result.socket.on('data', data => {
    console.log(data.toString());
  });
  console.log(result.socket);

}

test();
```

```
Socket {connecting: false, _hadError: false, _handle: TCP, _parent: null, _host: null, …}
index.ts:26
220 mx.google.com ESMTP 13si4866053otc.495 - gsmtp
index.ts:24
```

**Comment by srgl** [comment]
@JoshGlazebrook Ok, look, using this code I get random results, usually 9 out of 20. It depends on whether the line from the server comes with socks final handshake or not.
```
let {SocksClient} = require('socks')

const options = {
  proxy: {
    ipaddress: '127.0.0.1',
    port: 1080,
    type: 5
  },
  command: 'connect',
  destination: {
    host: 'gmail-smtp-in.l.google.com',
    port: 25
  }
};

(async function test () {
  let results = 0

  for (let i = 0; i < 20; i++) {
    let info = SocksClient.createConnection(options, (err, info) => {
      info.socket.on('data', data => {
        console.log(data.toString())
        results += 1
      })
    })
  }
  setTimeout(() => {
    console.log(`${results}/20`)
  }, 3000)
})()
```

**Comment by JoshGlazebrook** [comment]
This is going to be a pretty annoying fix lol.

Quick fix is here: https://github.com/JoshGlazebrook/socks/pull/26/files

Essentially I'll need to re-emit any excess data received in final handshakes/incoming connections (anything that emits the 'established' event) on the socket itself.

It also makes me wonder if I want to buffer the incoming data for the socks protocol handshake even though the data is so small (10 bytes-ish or less per packet), it may come in, in pieces. 

I'll also have to test if this quick fix works properly if there is a constant stream of incoming data that starts in the final handshake packet.

**Comment by srgl** [comment]
Yes, I think it looks 'annoying' because of the poor EventEmitter interface of net.Socket. On the other hand there is a `.read` method, but it is not recommended way of handling incoming data. 
@JoshGlazebrook for now I am using heroku's fork of socksv5 packet https://github.com/heroku/socksv5 and it works really good so far. So don't hurry, make it good. And the weird thing is that this packet is used in Nodemailer, which also works with smtp, and it seems there is no one who encountered the same issue.

**Comment by JoshGlazebrook** [comment]
This was fixed by #26 and is published to npm at 2.1.5
